File: proposed-projects/round-trip-experiment/reverse/20260909T124157Z_4ee67ba2/20260910T071826Z_7d46923a/ported_implementation/gbnf/grammar_graph/get_input_as_code_points.py
from typing import List

from .grammar_graph_types import ValidInput


def get_code_point(char: str) -> int:
    if not char:
        raise Exception(f"Could not get code point for character: {char}")
    code_point = ord(char[0])
    if not isinstance(code_point, int):
        raise Exception("code_point must be an integer!")
    return code_point


def get_input_as_code_points(src: ValidInput) -> List[int]:
    if isinstance(src, bool):
        raise Exception(f"Invalid input type: {type(src)}")

    if isinstance(src, int):
        return [src]

    if isinstance(src, (list, tuple)):
        for c in src:
            if not isinstance(c, int) or isinstance(c, bool):
                raise Exception(
                    f"code_point must be an integer for {c} if src is a list"
                )
        return list(src)

    if isinstance(src, str):
        return [get_code_point(char) for char in src]

    raise Exception(f"Invalid input type: {type(src)}")

File: packages/gbnf-experiment/data/20260909T180307Z_280f3890/ported_implementation/src/grammar-graph/get-input-as-code-points.ts
import type { ValidInput } from './grammar-graph-types.js';

export const getCodePoint = (char: string): number => {
  const codePoint = char.codePointAt(0);
  if (codePoint === undefined) {
    throw new Error(`Could not get code point for character: ${char}`);
  }
  if (!Number.isInteger(codePoint)) {
    throw new Error('code_point must be an integer!');
  }
  return codePoint;
};

export const getInputAsCodePoints = (src: ValidInput): number[] => {
  if (typeof src === 'number') {
    return [src];
  }

  if (Array.isArray(src)) {
    for (const c of src) {
      if (!Number.isInteger(c)) {
        throw new Error(`code_point must be an integer for ${c} if src is a list`);
      }
    }
    return src;
  }

  if (typeof src === 'string') {
    // Iterating a string with `Array.from` walks code points, matching Python's
    // per-character iteration rather than JS's UTF-16 code unit indexing.
    return Array.from(src).map(getCodePoint);
  }

  throw new Error(`Invalid input type: ${typeof src}`);
};
